### ISSUE TITLE
Transforms need to handle list of channel-first data

### ISSUE THREAD

**Issue by Nic-Ma** [Project-MONAI/MONAI#2220]
**Is your feature request related to a problem? Please describe.**
Currently, if using the default collate_fn, the `interted` data is a list of channel-first Tensor or numpy array, if we want to apply transforms on the data, most of the MONAI transforms can't support it.
For example, users want to use `CopyItemsd`, `AsDiscreted`, `KeepLargestConnectedComponentd`, etc.
All these transforms can only support **a batch-first Tensor** or **a channel-first numpy array**.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @rijobro and @wyli ,

Do you have any good ideas to handle this problem?
It would be very useful to create a post transform chain with `Invertd` transform, like:
```py
post_transforms = Compose([
    Activations(keys="pred"),
    CopyItemsd(keys="pred", times=1, names="inverted_pred"),
    Inverted(keys="inverted_pred", nearest_interp=False),
    AsDiscreted(keys=["pred", "inverted_pred"]),
    SaveImaged(keys=["pred", "inverted_pred"]),
])
```

Thanks.

**Comment by rijobro** [comment]
We could simply add an extra argument to their constructors, something like, `contains_batch_dim: bool = True`. We could potentially store that information in the `Transform` base class. In the majority of cases, that extra information wouldn't be used, but it means that in the future we could make other transforms compatible with both batch- and channel-first data.

What do you think?

**Comment by Nic-Ma** [comment]
Hi @rijobro ,

Thanks for your suggestion, I see your point.
So we still need to modify every post / utility transform to support a list of Tensor? Can we make some overall utility or optimize the `key_iterator()` to automatically support it?

Thanks.

**Comment by rijobro** [comment]
If used in a Compose object, I don't think it being a list of tensors is a problem, see [here](https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/transform.py#L47-L49). It will loop over each element, perform the transform on each of them and return a list.

The problem (as far as I understand) is that our post-processing transforms can't take CHW[D], and our pre-processing transforms can't take BCHW[D]. I think the solution I proposed above would try to harmonise these two. Ideally, we would remove the concept of pre- and post-processing transforms and all of our transforms would work on data that include the batch dimension and not, making them less prescriptive (e.g., someone might want to use `AsDiscreted` in their pre-processing pipeline).

**Comment by Nic-Ma** [comment]
Hi @rijobro ,

Thanks for your explanation.
The problem here is different from the `loop` solution in pre-transforms, here we handle a big dict, some inverted keys contain a list of Tensor(channel-first) while some keys contain just a Tensor(batch-first), it's **dict of list**.
But in the pre-transforms case, it's **list of dict**, we copied items when cropping, so we can use `loop` to handle every dict of the list directly.
I agree with you to add a data_shape property in the meta_data in future and then enhance transforms to support more flexible shape types.

Thanks.

**Comment by rijobro** [comment]
Hi @Nic-Ma , sorry I'm still struggling a little to visualise all the lists and dicts! Would you be able to give me a minimum working example? Is it something that the decollator could solve?

**Comment by Nic-Ma** [comment]
Hi @rijobro ,

Could you please help take a look at this draft PR: https://github.com/Project-MONAI/MONAI/pull/2244
I used `Activationd` transform as an example to show how to support a list of tensor.
And let me give you an example for the difference of the outputs between `RandCropByPosNegLabel` and `Invertd`:
1. For `RandCropByPosNegLabel`, if the input dict is:
```py
{"image": [1, 2, 3], "label": [1, 2, 3], "image_name": "testimg", "label_name": "testlabel", "extra": [4, 5, 6]}
```
After `RandCropByPosNegLabel(keys=["image", "label"], num_samples=2)`, the output is a **list of dicts**:
```py
[
    {"image": [1, 2], "label": [1, 2], "image_name": "testimg", "label_name": "testlabel", "extra": [4, 5, 6]},
    {"image": [2, 3], "label": [2, 3], "image_name": "testimg", "label_name": "testlabel", "extra": [4, 5, 6]}
]
```
So we can apply the following transforms for every item of the list, just same as other regular dict data.
2. But for the `Invertd` transform, after `Invertd(keys="image")`, the output is **still a dict, but only `image` became a list**:
```py
{"image": list(1, 2, 3), "label": [1, 2], "image_name": "testimg", "label_name": "testlabel", "extra": [4, 5, 6]}
```
So if we apply following transforms on this dict, the transforms should support both `Tensor`(label) and `list of Tensor`(image).

Thanks in advance.

**Comment by wyli** [comment]
The data after transform invention can be in different spatial sizes, just like the original image inputs. we don't have the batch dim for most of the preprocessing transforms, not sure why we need the concept of minibatch in the postprocessing here. minibatch is mostly useful for the stochastic gradient descent methods. if it's for performance purposes, I believe we have better ways to accelerate the postprocessing.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

PyTorch DataLoader collates batch in the program, but if want to remove the batch dim after model forward, we must ask users to manually decollate all the data before `post transforms` and `metrics`, right?
Something like:
```py
data = [{"image": img, "label": label} for img, label in zip(images, labels)]
metric_sum = 0
metric_count = 0
pre_transform = Compose([...])
post_transform = Compose([...])
metric = MeanDice()
dataset = Dataset(data=pre_transform)
dataloader = DataLoader(dataset=dataset, batch_size=8)

for d in dataloader:
    d["pred"] = net(d["image"])
    d["loss"] = loss(d["pred"], d["label"])
    list_d = decollate(d)  # decollate the batch data into a list of dict
    for i in range(len(list_d)):  # for-loop to run post transforms and metrics, instead of current batch operations
        list_d[i] = post_transform(list_d[i])
        metric_sum += metric(list_d[i]["pred"], list_d[i]["label"])
        metric_count += 1
```
I feel it's a big breaking change for our current post transforms, metrics and examples & tutorials, a little bit risky.
@ericspod  @rijobro  What do you guys think?

Thanks.

**Comment by wyli** [comment]
Sure, in the current codebase, the postprocessing s such as 'SaveImage' and 'InvertD' transform s already implicitly decollate batches. We should make the decollating explicit at the lowest level, explicit is better than implicit. if the user don't want the details they should use the postprocessing event handler.

**Comment by Nic-Ma** [comment]
Hi @ericspod and @rijobro ,

I think @wyli 's idea is to call the `decollate_batch` API after model forward(), then change all the following logic to work on a list of dict, or list of array. every item is channel-first. Just like above example code I prepared.
The benefit is to unify all the transforms to work for unique data shape: [C, H, W, [D]].
The drawback maybe we can't support ignite metrics anymore, and I am not sure we should do `decollate_batch` in which step of a `iteration`? Take the current MONAI supervised trainer as example:
```py
    def _iteration(self, engine: Engine, batchdata: Dict[str, torch.Tensor]):
        if batchdata is None:
            raise ValueError("Must provide batch data for current iteration.")
        inputs, targets = self.prepare_batch(batchdata, engine.state.device, engine.non_blocking)
        engine.state.output = {Keys.IMAGE: inputs, Keys.LABEL: targets}

        self.network.train()
        self.optimizer.zero_grad()
        engine.state.output[Keys.PRED] = self.inferer(inputs, self.network, *args, **kwargs)
        engine.fire_event(IterationEvents.FORWARD_COMPLETED)

        engine.state.output[Keys.LOSS] = self.loss_function(engine.state.output[Keys.PRED], targets).mean()
        engine.fire_event(IterationEvents.LOSS_COMPLETED)

        engine.state.output[Keys.LOSS].backward()
        engine.fire_event(IterationEvents.BACKWARD_COMPLETED)

        self.optimizer.step()
        engine.fire_event(IterationEvents.MODEL_COMPLETED)

        return engine.state.output
```
What do you guys think?
Or any other ideas?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

As I know, ignite metrics expect the data shape with batch dim, may I know some design reason for that?
We are investigating to decollate batch data after network forward() and loss forward(), then all the following post transforms and metrics will handle channel-first tensor in a list.
Because after inverting transforms, the data may have different init shape, so can't batch data anymore, and in this way, the post transforms can be designed the same as pre transforms to only support channel-first.

thanks.

**Comment by vfdev-5** [comment]
Hi @Nic-Ma ,

> As I know, ignite metrics expect the data shape with batch dim, may I know some design reason for that?

In our practice it the most common format for the models output and targets: `(B, C, ...)` and `(B, ...)`.
Also it helps to accelerate internal computations.

As I understand you would like to handle additionally two lists instead of tensors ? 
```python
metric.update((
    [ypred0, ypred1, ypred2, ...],
    [y0, y1, y2, ...],
))
```
where ypredX and yX have shapes `[C, ...]` and `[...]` or yX can be even a scalar.
I thinks this could be a feature for ignite...

**Comment by vfdev-5** [comment]
@Nic-Ma seeing your above example https://github.com/Project-MONAI/MONAI/issues/2220#issuecomment-848417625
```python
    for i in range(len(list_d)):  # for-loop to run post transforms and metrics, instead of current batch operations
        list_d[i] = post_transform(list_d[i])
        metric_sum += metric(list_d[i]["pred"], list_d[i]["label"])
        metric_count += 1
```
Seems like `list_d[i]["pred"]` and `list_d[i]["label"]` are 3D tensors which still could be used with current metrics if we were doing `.unsqueeze(0)` on them (which should be almost zero-cost). So, basically we can do something like that:
```python
    for i in range(len(list_d)):  # for-loop to run post transforms and metrics, instead of current batch operations
        list_d[i] = post_transform(list_d[i])
        metric.update(
            (list_d[i]["pred"].unsqueeze(0), list_d[i]["label"].unsqueeze(0))
        )
```
Please, correct me if I do not right get your idea.
Thanks

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

You are right, no need to totally change ignite metrics, but may need your help to add some wrapper layer in ignite to support a list of channel-first tensors or support 1 channel-first tensor.
We usually use ignite metrics in MONAI workflows directly, like this `Accuracy`:
https://github.com/Project-MONAI/tutorials/blob/master/modules/engines/unet_evaluation_dict.py#L111
Actually, I already added support for list of channel-first tensors in MONAI metrics:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/handlers/iteration_metric.py#L91

Thanks.
